Quantum-Control-Applications/Superconducting/Multi Flux Tunable transmon/Use Case 1 - Two qubit gate optimization with cryoscope/readme.md

```diff
@@ -0,0 +1,192 @@
+ The possibly distorted response can then be pre-corrected by designing suitable IIR and FIR filters.
+
+This experiment and implementation in QUA has already been documented in the following [use-case](#2).
+The main difference here is that a constant amplitude flux pulse has been used to measure the step response of the flux line.
+
+
+### 3.2 [IIR and FIR filter implementation](cryoscope.py)
+
+The method used to design the filters is to fit the flux step response with one or several exponential decay functions 
+and use analytical formulas to derive the corresponding FIR and IIR taps. 
+This exponential behaviour is characteristic of the effect of a bias-T on the flux line.
+
+These functions can be found in [filter_functions.py](filter_functions.py) and their usage is shown in the snippet below.
+
+```python
+import scipy.optimize
+from filter_functions import expdecay, filter_calc
+## Fit step response with exponential
+[A_exp, tau_exp], _ = scipy.optimize.curve_fit(expdecay, xplot, step_response)
+print(f"A: {A_exp}\ntau: {tau_exp}")
+
+## Derive IIR and FIR corrections
+fir, iir = filter_calc(exponential=[(A, tau)])
+print(f"FIR: {fir}\nIIR: {iir}")
+```
+Once the filter parameters have been found, the configuration needs to be updated accordingly.
+An example is shown on the snippet below 
+and more details about the implementation of FIR and IIR filters in QUA can be found [here](https://qm-docs.qualang.io/guides/output_filter#id1).
+```python
+"controllers": {
+    "con1": {
+        "analog_outputs": {
+            1: {"offset": -0.0102 * 1}, # q0 I
+            2: {"offset": 0.0316 * 1},  # q0 Q
+            3: {"offset": 0.0},  # q1 I
+            4: {"offset": 0.0},  # q1 Q
+            5: {"offset": 0.0},  # resonators I
+            6: {"offset": 0.0},  # resonators Q
+            7: {"offset": offset7, "filter": {"feedforward": fir, "feedback": iir}},   # qo flux line
+            8: {"offset": offset8, "filter": {"feedforward": fir1, "feedback": iir1}}, # q1 flux line
+        },
+    },
+},
+```
+The figure below shows the results of the filter implementation.
+
+The top-left figure represents the step response of the flux line and the crorresponding exponential fit.
+The top-right figure represents the filters correction as well as the calculated taps.
+The bottom-left plot is the normalized flux pulse without and with filter and the bottom-right figure is a zoom with the $\pm$ 1% region.
+
+![results](Pulse_response.png)
+
+
+### 3.3 [Two qubit SWAP spectroscopy](CZ02_11.py)
+
+In order to characterize the quality of the corrections, a two qubit SWAP spectroscopy is implemented.
+It consists of four steps:
+1. Prepare the qubits in the ground state. For simplicity, this is done by waiting a cooldown time equivalent to 5*T1, but an active reset protocol can also be used to gain in efficiency.

```

Comment:
Why are we not doing it? It would definitely look better...

Reply:
I think it was for simplicity since Kevin had only small time windows to work on this...